docs/tutorials/go-built-in.md

```diff
@@ -103,579 +101,527 @@ func NewKVStoreApplication() *KVStoreApplication {
+will be delivered to the application.
+
+To implement these calls in our application we're going to make use of Badger's 
+transaction mechanism. 
+
+First, let's create a new Badger `Txn` during `BeginBlock`:
 
 ```go
 func (app *KVStoreApplication) BeginBlock(req abcitypes.RequestBeginBlock) abcitypes.ResponseBeginBlock {
- app.currentBatch = app.db.NewTransaction(true)
- return abcitypes.ResponseBeginBlock{}
+	app.currentBatch = app.db.NewTransaction(true)
+	return abcitypes.ResponseBeginBlock{}
 }
-
 ```
 
-Here we create a batch, which will store block's transactions.
+Next, let's modify `DeliverTx` to add the `key` and `value` to this `Txn` every time our application
+receives a new `RequestDeliverTx`.
 
 ```go
 func (app *KVStoreApplication) DeliverTx(req abcitypes.RequestDeliverTx) abcitypes.ResponseDeliverTx {
- code := app.isValid(req.Tx)
- if code != 0 {
-  return abcitypes.ResponseDeliverTx{Code: code}
- }
+	if code := app.validateTx(req.Tx); code != 0 {
+		return abcitypes.ResponseDeliverTx{Code: code}
+	}
 
- parts := bytes.Split(req.Tx, []byte("="))
- key, value := parts[0], parts[1]
+	parts := bytes.Split(req.Tx, []byte("="))
+	key, value := parts[0], parts[1]
 
- err := app.currentBatch.Set(key, value)
- if err != nil {
-  panic(err)
- }
+	err := app.currentBatch.Set(key, value)
+	if err != nil {
+		panic(err)
+	}
 
- return abcitypes.ResponseDeliverTx{Code: 0}
+	return abcitypes.ResponseDeliverTx{Code: 0}
 }
 ```
+Note that we check the validity of the transaction _again_ during `DeliverTx`.
+Transactions are not guaranteed to be valid when they are delivered to to an
+application. 
 
-If the transaction is badly formatted or the same key=value already exist, we
-again return the non-zero code. Otherwise, we add it to the current batch.
+Also note that we don't commit the Badger `Txn` we are building during `DeliverTx`. 
+Other methods, such as `Query`, rely on a consistent view of the application's state.
+The application should only update its state when the full block has been delivered.
 
-In the current design, a block can include incorrect transactions (those who
-passed CheckTx, but failed DeliverTx or transactions included by the proposer
-directly). This is done for performance reasons.
+The `Commit` method indicates that the full block has been delivered. During `Commit`,
+the application should persist the pending `Txn`.
 
-Note we can't commit transactions inside the `DeliverTx` because in such case
-`Query`, which may be called in parallel, will return inconsistent data (i.e.
-it will report that some value already exist even when the actual block was not
-yet committed).
-
-`Commit` instructs the application to persist the new state.
+Let's modify our `Commit` method to persist the new state to the database:
 
 ```go
 func (app *KVStoreApplication) Commit() abcitypes.ResponseCommit {
- app.currentBatch.Commit()
- return abcitypes.ResponseCommit{Data: []byte{}}
+	app.currentBatch.Commit()
```

Comment:
This can fail, and we should check.

In particular: It is possible the dbtx has conflicts.